balanced-optiml/bal-optiml-library/library.whizzml

```diff
@@ -0,0 +1,139 @@
+;; Transposes the matrix: [[a b c] [d e f]] to [[a d] [b e] [c f]]
+(define (_transpose m)
+  (apply map list m))
+
+;; Deletes resources ignoring errors
+(define (_safe-delete id)
+  (try (delete id)
+       (catch e
+         (log-info (str "Error deleting resource " id " ignored:" e))
+         false)))
+
+;; Check that the objective field exists, is a categorical field and
+;; contains the rare class
+(define (_check-cat fields objective-field rare-class)
+  (let (field (find-field fields objective-field))
+    (when (not (find-field fields objective-field))
+      (raise {"message" (str "The objective field "
+                             objective-field
+                             " cannot be found in this dataset")
+              "code" 100}))
+    (when (not (categorical-field? field))
+      (raise {"message" (str "The objective field "
+                             objective-field
+                             " is not a categorical field."
+                             " Only categorical fields can be used.")
+              "code" 101}))
+    (when (not (member? rare-class (field-categories field)))
+      (raise {"message" (str "The objective field "
+                             objective-field
+                             " does not contain the "
+                             rare-class
+                             " category. Please check its name.")
+              "code" 102}))))
+
+;; Function that splits a dataset in training, validation and test datasets
+(define (_create-train-validation-test ds-id train-rate seed)
+  (let (first-split (create-dataset-split ds-id train-rate seed)
+        second-split (create-dataset-split (last first-split) 0.5 seed))
+    (cons (head first-split) second-split)))
+
+
+;; Function to apply stratified split for the rare class of the objective field
+;; with a given train
+(define (rare-stratified-split-fn ds-id training-rate objective-field
+                                  rare-class project)
+  (let (dataset (fetch ds-id)
+        ds-name (dataset "name")
+        fields (resource-fields dataset)
+        _ (_check-cat fields objective-field rare-class)
+        objective-id ((find-field fields objective-field) "id")
+        filters [(flatline "(= (f {{objective-id}}) {{rare-class}})")
+                 (flatline "(!= (f {{objective-id}}) {{rare-class}})")]
+        names [rare-class (str "not " rare-class)]
+        cat-ds-ids (iterate (acc [] expr filters name names)
+                     (append acc
```

Comment:
iterate + append is just map, right? although i like map better, it's okay to keep iterate if you prefer, but then perhaps you could also use it in the computation of ds-ids below for consistency... seeing both used next to the other made me double check that iterate wasn't doing something else.

Reply:
ok, changing to map